Fix blocking worker threads, possible race conditions and improve logging

mutax opened the pull request:
With this PR we modify the scheduling of jobs to the worker pool, improve logging and allow single unit tests to be executed.

On the way we fixed several bugs and inconsistencies in the existing code, required to implement the changes.

Please refer to the commit message of the individual changes, the explanation for the changes in the scheduling is rather long.

mutax committed:
[tests] allow running only a subset of unit tests

mutax committed:
[tests] replace assertEquals for membership test with assertIn

mutax committed:
[cli] print the invalid command for debugging the unittests

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -97,7 +102,8 @@ def __lt__(self, other):
     def __hash__(self):
-        return hash(self.__repr__())
+        # with the original repr this is broken, as __eq__ takes the fn into account as well!
+        return hash((self.idn, self.fn))
```
Can we use ``self.identifier`` here? So `self.fn.__name__` instead of `self.fn`.

sven-rosenzweig replied:
Do we need to override ``__hahs__``? 
According to the docs: 
Called by built-in function [hash()](https://docs.python.org/2/library/functions.html#hash) and for operations on members of hashed collections including [set](https://docs.python.org/2/library/stdtypes.html#set), [frozenset](https://docs.python.org/2/library/stdtypes.html#frozenset), and [dict](https://docs.python.org/2/library/stdtypes.html#dict). [__hash__()](https://docs.python.org/2/reference/datamodel.html#object.__hash__) should return an integer

We simply add Runnable in a Queue (managing a list), right?

mutax replied:
`__hash__` was overwritten upstream, but buggy -- if two objects compare equal, they should also have identical hash values and vice-versa.

After fixing that I was able to use a ` job: Runnable` as keys in my dictionaries for the rerunner and stats.

Similarily we have to use `fn` and not just `fn.__name__` , because the name does not include the package name and in the `__eq__`  method fn is compared and not the __name__.

 Hence `__eq__` would be false, but hash() would return the same hash if a function from another python package is used, e.g. to overwrite the callback.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -149,15 +149,29 @@ class UniqPriorityQueue(eventlet.queue.Queue):
 
-    def _get(self, heappop=heapq.heappop):
-        return heappop(self.queue)
+        try:
```
Do we really need this? 
As the passive queue is the only UniqPriorityQueue left. As for the current implementation all jobs added to the passive queue have the same priority. 

The active queue becomes a FiFo Queue, priority won't matter anymore.

mutax replied:
We should be able to use a UniqFifoQueue for the passive queue as well now, that is correct, because we handle the only 2 priorities used in two queues now. As long as we do not need more priorities that will work. 

While in principle I am open for that I would do that change separately in case we find out we would benefit from actually using more priorities to, e.g. , first delete ports and then create new ones/call update.

mutax replied:
do we want to do that or keep it as is for now?

sven-rosenzweig replied:
As discussed, we keep this for now because we might need it in future again.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -173,6 +174,93 @@ def _put(self, item):
         return heapq.heappop(self.queue)
 
+class JobRerunner():
+    """ Threadsave data structure to reschedule jobs when already running
```
Good explanation. 
Maybe we could emphasise, that a job is runs at max twice as long as it is running. So it does not matter if we increase the counter to 100 or 2.

mutax replied:
Currently there is one comment addressing that, should we make it more obvious?
 https://github.com/sapcc/networking-nsx-t/blob/12ae5a49f44baaf2f26ebbcb1c30876a3b7fbd80/networking_nsxv3/common/synchronization.py#L236
 _Note that a job will intentionally only be re-executed once, independently of how often it was added while a worker was already running it._

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
Use a const here? it is also used in the ``add_job`` function.

mutax replied:
done, but not sure if that made the code more readable, though

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -232,10 +321,29 @@ def _start(self):
-                self._workers.spawn(job.fn, job.idn)#.wait()
-                self._active.task_done()
+                job = self._rerunner.pop()
+                from_queue = False
```
Would rename it to ``pulled_from_active_queue``

mutax replied:
I skipped active in the name because I found it too long.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -173,6 +174,93 @@ def _put(self, item):
+    """
+
+    def __init__(self):
+        self._ready = collections.deque()
```
this is the queue storing jobs which needs to be rerun? Would change the naming here.

sven-rosenzweig replied:
Also the job ``def pop(self)`` is misleading.

mutax replied:
renamed

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -72,11 +72,66 @@ def __init__(self, idn, fn, priority=Priority.LOWEST):
+
+        scheduled = '-'
+        started = '-'
+        jobdone = '-'
```
I would ``jobdone`` to ``runtime``. the label in the format string says so: ``runtime: {jobdone}``.

mutax replied:
right, forgot to rename the var when I added started and the calls in the wrapper - will do!

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -222,10 +278,13 @@ def pop(self) -> Runnable:
             try:
                 job = self._ready.popleft()
+                job.set_rescheduled()
                 LOG.debug("JobRerunner had rerunnable job: %s", job)
```
Do we need this debugging log?

mutax replied:
We are currently running with INFO so it does not hurt in case we need to debug the changed scheduling -- would keep it until we are sure we don't need it.

sven-rosenzweig replied:
But the same info is already contained in the info log (job is rerun, id and function name)?

mutax replied:
the ui didn't show the next line in the conversation pane, did not see the info below. I reworded the log messages of the info message to be consistent, also the statusline method uses self in the Runnable, so all log messages will include the same pattern in case we modify repr later. debug removed.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -236,17 +295,21 @@ def job_done(self, job: Runnable):
             if count == 1:
                 del self._running[job]
                 LOG.debug("JobRerunner job %s is done, no reruns requested", job)
```
Do we need this debugging log?

mutax replied:
We are currently running with INFO so it does not hurt in case we need to debug the changed scheduling -- would keep it until we are sure we don't need it.

mutax replied:
the ui didn't show the next line in the conversation pane, did not see the info below. I reworded the log messages of the info message to be consistent, also the statusline method uses self in the Runnable, so all log messages will include the same pattern in case we modify repr later. debug removed.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -260,25 +323,26 @@ def add_job(self, job: Runnable) -> bool:
+
+        # let's log these as info for debugging, they should be sufficient in prod
+        # to find issues with the JobRerunner:
+        LOG.info("JobRerunner stat: %d jobs waiting, total submission count: %d", len(self._running), sum)
```
What insights do we get with the total count? 
Does it make sense to replace ``%d jobs waiting`` with ``%d jobs running``  as we take the infromation from ``len(self._running)``?

mutax replied:
The total count gives us a metric to compare to the 40 worker threads, as that gives us an indication how big the problem we are trying to fix (all workers blocked with the same requests) is in prod. If we grep the log and see this number increasing without the number of jobs increasing, we know that we would have blocked all workers with the current code.

mutax replied:
jobs waiting was meant to say "jobs waiting for reexecution" but you are right, that is wrong as not all of them are waiting for that.  Jobs running should be more or less correct, although this is just the number of jobs tracked in this class and not directly coupled to the worker pool. I think I will rename it to 'jobs tracked' -- thats less confusing

sven-rosenzweig replied:
I like 'jobs tracked'.

mutax committed:
Reworking scheduling of jobs to runners

The main change in this commit is a modification of the way jobs are
handled in the NSX-T agent. Please see the JobRerunner class for an
in-depth explanation of the changes.

Before this commit jobs are added to one of two queues, called active
and passive. The active queue contains all requests coming in via API
calls, while the passive queue is filled with maintenance and resync
jobs. Both queues used to be priority queues allowing each element to be
added only once.

Jobs then were taken from the active queue until empty, then jobs from
the passive queue would be added to the active queue.

Jobs taken from the active queue would then be submitted to a worker
pool allowing up to 40 greenthreads to run the jobs concurrently

However, to avoid race conditions, only one job is allowed to run per
OpenStack-ID. If more than one job is scheduled to run on the pool,
these additional jobs will wait on a lock and block the worker thread
until the first job is done.

This means that the agent can be blocked and appear fully occupied,
handling 40 tasks simultaneously, while in reality most or all tasks are
waiting for each other.

Instead of scheduling all jobs to the worker pool immediately, risking
a lock, we now first check if the same job is already running, and if
this is the case we will rerun the job after it has finished.
We then can schedule another job that can run to the worker instead.

We need to rerun the job, because a jobs can run for several seconds and
new API requests could arrive during that time.

With this change we also prevent rerunning the same job more than once,
when additional requests arrive while the job is already marked for
re-execution.

While implementing these changes, we found out that some api calls,
and thus the resulting jobs, will get a dictionary and not a string as
parameter, although indicated differently in the code.

To support these calls, we have to handle that case as well, they are:
  address_group_update and {enable, disable, update}_policy_logging

Additional fixes and enhancements in this commit:

UniqPriorityQueue:

- fixing add()

If a job is about to be added a second time, but with a higher priority,
the job will correctly not be added, but the priority of the existing
job was not updated. This means jobs from the passive queue, that have a
lower priority, will always be executed last, even if a high-priority
job arrived via API call.

We changed the active queue to a Fifo, to prevent passive jobs to never
get executed and keep execution order of api calls if possible.
With the fix in place, however, we can switch back to a prio queue if
needed.

Runnable:

- fix hash() and make repr more verbose

The Runnable class was not following the requirements for objects that
compare equal to also have the same hashvalue. Also the Runnable was
only taking the OpenStack ID into account, not the name of the function.
Thus a Runnable could, e.g., not be used correctly as a key in
dictionaries.

- __repr__

repr was updated to include the name of the function,
so we see what kind of update is being executed in the logs.

- __lt__

making Runnable order items with same priority by age, preventing
jobs  from overtaking each other.

- add timing info for logging

We currently do not get good info about the timings or basic stats
of the jobs running. This commit adds timing info to Runnable and
a method to extract them as string for logging.

felix-kaestner commented on networking_nsxv3/tests/unit/realization/test_coordination.py:
```diff
@@ -0,0 +1,182 @@
+        self.assertNotIn('runtime: -', job.get_statline())
+
+        callback.assert_called()
+
+
+
+
+
```
```suggestion
```

joker-at-work commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -90,25 +151,204 @@ def __ne__(self, other):
+
+    When a job is retrieved from the active queue and already running in a worker thread,
+    there is a chance that another job for the same object is added to the active
+    queue and also started in a worker. While there then is some locking happening
+    that should prevent race conditions, this still blocks the worker thread(s),
```
minor: "While there then is some locking happening" sounds strange

mutax replied:
rewritten.

joker-at-work commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -90,25 +151,204 @@ def __ne__(self, other):
+                    self._to_rerun.append(job)
+            else:
+                # prevent the error from spreading
+                del self._running[job]
```
This could raise `KeyError` if the `job`is not in there anymore. We can still get `count == 0`, because of `self._running.get()` in line 300.

joker-at-work commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -90,25 +151,204 @@ def __ne__(self, other):
+                # if the counter is < 0 our accounting is wrong,
+                # so we fix it and run the job.
+                self._running[job] = 1
+                LOG.debug("JobRerunner no identical job is currently running, can start %s", job)
```
Why no `job.get_statline()`?

mutax replied:
the statusline is relatively verbose and the stats are only really interesting when a state change occurs, e.g. when a job is started or done or rescheduled. Its sufficient to get the identifier when debugging the flow of execution, I think.

joker-at-work commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -90,25 +151,204 @@ def __ne__(self, other):
+                return True
+
+            self._running[job] = count + 1
+            LOG.debug("JobRerunner job %s already running, marked for rescheduling, count: %d ", job, count)
```
Why no `job.get_statline()`?

mutax replied:
Same as above, here we are only really interested in how often a job has been re-requested, so no need for all the details

joker-at-work commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -90,25 +151,204 @@ def __ne__(self, other):
+            sum = 0
+            for job, scount in self._running.items():
+                sum += scount
+                LOG.debug("JobRerunner stat: job %s is running, submission count: %d", job, scount)
```
Why no `job.get_statline()`?

mutax replied:
see above

joker-at-work commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -90,25 +151,204 @@ def __ne__(self, other):
+            self._running[job] = count + 1
+            LOG.debug("JobRerunner job %s already running, marked for rescheduling, count: %d ", job, count)
+
+            sum = 0
+            for job, scount in self._running.items():
+                sum += scount
+                LOG.debug("JobRerunner stat: job %s is running, submission count: %d", job, scount)
```
On every submission we print the stats? Don't you want to mark this as "Please remove once we're ready for prod"? Seems quite excessive. How large do you think `self._running`? will become?

mutax replied:
max 40, as we have 40 workers running. If all workers are busy, the loop adding new jobs will not add more jobs.

joker-at-work replied:
So that could be 40 lines per ... second per BB? Every time we try to run a job hm ...

mutax replied:
only when using debug, which we usually only do in qa - so I think in those cases its okay?

joker-at-work replied:
Ah, I thought we'd be running at debug-level, but we're on INFO, you're right.

joker-at-work commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -169,10 +410,31 @@ def _start(self):
+                    # ideally we would be able to add a callback to the
+                    # greenthread, but this is hidden in the pool, so
+                    # let's wrap the function once more.
+                    def wrap(rerun, ajob):
+                        ajob.set_start()
+                        ajob.fn(ajob.idn)
+                        ajob.set_done()
+                        rerun.job_done(ajob)
```
Why does this not need a try-finally?

mutax replied:
was added and the whole logic moved into the runnable, so the set methods are no longer public and the runnable is under control of whats happening.

joker-at-work reviewed (commented):
first glance over it

mutax commented:
marked PR as draft, need some more work -- contradicting the code and comments in the code, jid is not always a string containing an OS-ID, sometimes its a dictionary with the parameters for the actual change.  This of course breaks a lot of assumptions.

felix-kaestner commented on networking_nsxv3/common/synchronization.py:
```suggestion
        return sum(count for count, _ in self._runnables)
```

felix-kaestner commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -129,18 +576,24 @@ class Runner(object):
                  workers_size=1):
         # if queue_size is < 0, the queue size is infinite.
-        self._active = UniqPriorityQueue(maxsize=active_size)
+        self._active = UniqFiFoQueue(active_size)
```
```suggestion
        self._active = UniqFiFoQueue(maxsize=active_size)
```

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -238,14 +273,179 @@ def _get(self):
+        In this datastructure we are keeping track of the jobs and their parameters for the JobRerunner,
+        based on the identifier of the jib, which is telling us the type of job (callback name) and openstack id.
+
+        There can be multiple similar jobs (same id, same method) but with different parameters, then parameter
```
Dicts only for ``address groups`` and ``logging``. Could mention this in this docstring.

mutax replied:
I found these:  `address_group_update `and `{enable, disable, update}_policy_logging` in AgentRealizer,  are those all?

sven-rosenzweig replied:
I guess so.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -238,14 +273,179 @@ def _get(self):
+        done will then choose a job from the list, that is supposed to run again,
+        remove it from the list and return it to the JobRerunner.
+
+        This might be the same job that was just finished or it could be a different one.
```
With different one == same job but other parameters, right?

mutax replied:
yes, will clarify

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
What happens if we raise this exception? Might a job not be added to the queue?

mutax replied:
Yes, this is a programming error and should never happen -- there is not much we can do at this point than raise an exception and log it as an error.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
Does not work as stated in the comment.

mutax replied:
fixed - please check, reworked the surrounding code.

sven-rosenzweig commented on networking_nsxv3/common/synchronization.py:
```diff
@@ -296,26 +496,35 @@ def get_rerunnable(self) -> Runnable:
 
-            if count == 1:
-                del self._running[job]
+            joblist: JobList = self._running[job.identifier]
```
```
            joblist: JobList = self._running[job.identifier]
            if joblist.empty():
                try:
                    del self._running[job.identifier]
                except KeyError:
                    # should never happen, we just fetched it above
                    pass
                LOG.info("JobRerunner (done, no reruns requested) %s", job.get_statline())
                return
            
            # TODO(mutax): make debug after nsx-t issues solved
            LOG.info("After job done %s", joblist)

            next_job = None
            try:
                next_job = joblist.done(job)
            except Exception as e:
                LOG.error("Error getting next job after a job was done: %s", e)

            if next_job: 
                LOG.info("JobRerunner (done, got rerun) done: %s next: %s", job.get_statline(), next_job.get_statline())
                with LockManager.get_lock(self._lname_torerun):
                    self._to_rerun.append(next_job)
            else:
              LOG.info("JobRerunner (done, no reruns requested) %s", job.get_statline())
```
just an alternative suggestion for this. not 100% sure if it covers your implementation.

mutax replied:
what is the advantage/what is your reasoning?

I am not sure if we could always guarantee to clean up properly with the suggested change, the way
I am doing it seems more straight forward and 'saver' -- if I don't miss your point here.

If we first update the counters (which is happening in the done call) and then check if the list is empty, we can always be sure that at that point the list entry gets cleaned up. If we first check if its empty and then update the bookkeeping, we would keep the list around even when it was the only job running.

sven-rosenzweig replied:
I tried to separate the case if ``joblist.empty() == true`` (so no more jobs to execute) and try to return early in this case. From my point of view this would be simpler to understand.

sven-rosenzweig replied:
my comment is meant as optional, so no need to change anything here.
